lib/libzfs/libzfs_sendrecv.c

```diff
@@ -4393,9 +4416,34 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 	redacted = DMU_GET_FEATUREFLAGS(drrb->drr_versioninfo) &
 	    DMU_BACKUP_FEATURE_REDACTED;
 
-	if (zfs_dataset_exists(hdl, name, ZFS_TYPE_DATASET)) {
+	if (flags->heal) {
+		if (flags->isprefix || flags->istail || flags->force ||
+		    flags->canmountoff || flags->resumable || flags->nomount ||
+		    flags->skipholds) {
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "corrective recv can not be used when combined with"
+			    " this flag"));
+			err = zfs_error(hdl, EZFS_INVALIDNAME, errbuf);
+			goto out;
+		}
+		uint64_t guid =
+		    get_snap_guid(hdl, name, strchr(destsnap, '@') + 1);
+		if (guid == 0) {
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "corrective recv must specify an existing snapshot"
+			    " to heal"));
```

Comment:
How does one know which snapshot(s) need healing from a set of errors like this: 

```errors: Permanent errors have been detected in the following files:

        <metadata>:<0x0>
        <metadata>:<0x1>
        <metadata>:<0x3f3>
        /rust/private/
        whirlpool/data:<0x0>
```

And would it be worthwhile adding some kind of instructions like that to the manpage change further down or does it belong elsewhere?

Reply:
`There are likely several snapshots referring to the same potentially corrupted data so there may be many snapshots with the above condition holding that are able to heal a single block.` this comment from the PR body partially answers my question. But in my case the only tangible thing being `/rust/private` had nothing in it prior to the most recent send/receive except other datasets and post that send/receive had one folder which seemed fine (did not do a diff -rq to check mind you). So I'm guessing that in my case given I have 3 metadata errors and one empty dataset error and one empty mount point error, there were no data blocks and this PR couldn't have helped me anyway, is that right?

Reply:
yup, this PR would not have healed your metadata errors